packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,334 @@
+      Subscription<IDecodedMessage>
+    >;
+
+    for (const item of subscriptions.values()) {
+      const values = map.get(item.pubSubTopic) || [];
+      const nextValues = item.decoders.map((decoder) => decoder.contentTopic);
+      map.set(item.pubSubTopic, [...values, ...nextValues]);
+    }
+
+    return map;
+  }
+
+  private onRequest(streamData: IncomingStreamData): void {
+    log("Receiving message push");
+    try {
+      pipe(streamData.stream, lp.decode(), async (source) => {
+        for await (const bytes of source) {
+          const response = FilterPushRpc.decode(bytes.slice());
+
+          const { pubsubTopic, wakuMessage } = response;
+
+          if (!wakuMessage) {
+            log("Received empty message");
+            return;
+          }
+
+          const subscription = Array.from(this.subscriptions.values()).find(
+            (s) =>
+              (s as Subscription<IDecodedMessage>).pubSubTopic === pubsubTopic
+          ) as Subscription<IDecodedMessage> | undefined;
+
+          if (!subscription) {
+            log(`No subscription locally registered for topic ${pubsubTopic}`);
+            return;
+          }
+
+          await this.pushMessage(subscription, wakuMessage);
+        }
+      }).then(
+        () => {
+          log("Receiving pipe closed.");
+        },
+        (e) => {
+          log("Error with receiving pipe", e);
+        }
+      );
+    } catch (e) {
+      log("Error decoding message", e);
+    }
+  }
+
+  private async pushMessage<T extends IDecodedMessage>(
+    subscription: Subscription<T>,
+    message: WakuMessage
+  ): Promise<void> {
+    const { decoders, callback, pubSubTopic } = subscription;
+
+    if (!decoders || !decoders.length) {
+      log(`No decoder registered`);
+      return;
+    }
+
+    const { contentTopic } = message;
+    if (!contentTopic) {
+      log("Message has no content topic, skipping");
+      return;
+    }
+
+    let didDecodeMsg = false;
+    // We don't want to wait for decoding failure, just attempt to decode
+    // all messages and do the call back on the one that works
+    // noinspection ES6MissingAwait
+    decoders.forEach(async (dec: IDecoder<T>) => {
+      if (didDecodeMsg) return;
+      const decoded = await dec.fromProtoObj(
+        pubSubTopic,
+        message as IProtoMessage
+      );
+      // const decoded = await dec.fromProtoObj(pubSubTopic, message);
+      if (!decoded) {
+        log("Not able to decode message");
+        return;
+      }
+      // This is just to prevent more decoding attempt
+      // TODO: Could be better if we were to abort promises
+      didDecodeMsg = Boolean(decoded);
+      await callback(decoded);
+    });
+  }
+
+  private async unsubscribe(
+    topic: string,
+    contentTopics: string[],
+    requestId: string,
+    peer: Peer
+  ): Promise<void> {
+    const unsubscribeRequest = FilterSubscribeRpc.createUnsubscribeRequest(
+      topic,
+      contentTopics,
+      requestId
```

Comment:
You are not meant to re-use a previous request id when unsubscribing. It should be a new request id.
The remote node remember your subscription thanks to your peer id.

Cc @jm-clius to confirm.

Reply:
Indeed. Request ID now _only_ ties a request to a response. For this reason they SHOULD be unique, to avoid ambiguity in responses. Remote peers keep state only against peer ID.